### DISCUSSION TITLE
v2.16.0

### DISCUSSION THREAD

**Discussion by sojan-official** [chatwoot/chatwoot#6917] [Releases & Announcements]
## Changelog

-  UI to view Audit Logs ( EE edition only )
- Context Menu for  messages (  while right-clicking  )
- Ability for admins to edit agent availability status
- Ability to link to a specific message
- Support for message-updated events in the telegram channel
- Ability to download heatmap reports
- support routing in the email channel using ex-original-to in the email header
- Additional automation filters ( email and phone number based )
- Support for sending multiple attachments to Whatsapp inbox
- Improved calculations of reporting metrics
- Advanced IndexDB based caching for common objects like inboxes, labels etc
- Improved contact import
- Numerous bug fixes and enhancements
- New Languages: Icelandic

###
Thanks to @ChamathKB, @s00d, @jordan-brough @clairton, @balawa, @gitstart, @BikashSah999  for the contributions

<hr /><em>This discussion was created from the release <a href='https://github.com/chatwoot/chatwoot/releases/tag/v2.16.0'>v2.16.0</a>.</em>

**Comment by silvaartur** [comment]
Hi @sojan-official I think #6788 is important to be merged because this bug broke email functionality.

**Reply by sojan-official** [reply]
I will take it over and get it merged. Was waiting on the original contributor

**Comment by arladmin** [comment]
@sojan-official 

In the latest release, 2.16.0, the API call isn't able to go through.

Error:
![image](https://user-images.githubusercontent.com/64059154/232744624-f8de2fd4-19bd-4755-bd7b-ab42529e1445.png)

**Reply by sojan-official** [reply]
@scmmishra  could you take a look ?

**Reply by scmmishra** [reply]
@arladmin can you send the payload you are using?

**Reply by arladmin** [reply]
@scmmishra 

it's something like this:
![image](https://user-images.githubusercontent.com/64059154/232837012-acf860b8-5c43-4d81-99a2-a4acf53fc6fc.png)

This is how it's called via JS:
```
      window.$chatwoot.setConversationCustomAttributes({
          software_version,
          displayMode,
          languages,
          connection,
          userAgent,
          userAgentData,
          ipAddress,
          geo,
          uuid
      });
```

Although, not sure why 500 Error should be payload dependent.

**Reply by scmmishra** [reply]
@arladmin I wasn't able to replicate this so far, is this Chatwoot Cloud or your self-hosted instance? If it is self-hosted, is it possible to send an error log?

**Reply by arladmin** [reply]
@scmmishra 
Self-Hosted (Cloudron)

Here are the concerned logs:
```
Apr 19 10:43:54 I, [2023-04-19T05:13:54.506007 #14] INFO -- : [79cca148-c9ab-4d72-9ea2-986399ab7746] Processing by Api::V1::Widget::ContactsController#update as JSON
Apr 19 10:43:54 I, [2023-04-19T05:13:54.506084 #14] INFO -- : [79cca148-c9ab-4d72-9ea2-986399ab7746] Parameters: {"custom_attributes"=>{"uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}, "website_token"=>"[FILTERED]", "cw_conversation"=>"eyJhbGciOiJIUzI1NiJ9.eyJzb3VyY2VfaWQiOiI1OTExNjU1Mi1mM2U3LTQ1MjUtYmEyNS1jZDljZDAzM2Y2ODQiLCJpbmJveF9pZCI6MX0.tuf5tICzhLaYYxshOmub1KVsM_oTL09dH5PnS3h4EjU", "contact"=>{"custom_attributes"=>{"uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}}}
Apr 19 10:43:54 I, [2023-04-19T05:13:54.507942 #14] INFO -- : source=rack-timeout id=795bcf05-7080-4d7c-baeb-e07f0b4cdc7c timeout=15000ms state=ready
Apr 19 10:43:54 I, [2023-04-19T05:13:54.508924 #14] INFO -- : [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] Started POST "/api/v1/widget/conversations/set_custom_attributes?website_token=[FILTERED]&cw_conversation=eyJhbGciOiJIUzI1NiJ9.eyJzb3VyY2VfaWQiOiI1OTExNjU1Mi1mM2U3LTQ1MjUtYmEyNS1jZDljZDAzM2Y2ODQiLCJpbmJveF9pZCI6MX0.tuf5tICzhLaYYxshOmub1KVsM_oTL09dH5PnS3h4EjU" for 122.180.29.153 at 2023-04-19 05:13:54 +0000
Apr 19 10:43:54 I, [2023-04-19T05:13:54.514871 #14] INFO -- : [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] Processing by Api::V1::Widget::ConversationsController#set_custom_attributes as JSON
Apr 19 10:43:54 I, [2023-04-19T05:13:54.515051 #14] INFO -- : [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] Parameters: {"custom_attributes"=>{"software_version"=>"6.5", "displayMode"=>"browser tab", "languages"=>["en-GB"], "connection"=>{"downlink"=>3.95, "effectiveType"=>"4g", "rtt"=>100}, "userAgent"=>"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36", "userAgentData"=>{"brands"=>[{"brand"=>"Chromium", "version"=>"112"}, {"brand"=>"Google Chrome", "version"=>"112"}, {"brand"=>"Not:A-Brand", "version"=>"99"}], "mobile"=>false, "platform"=>"Windows"}, "ipAddress"=>"", "geo"=>{"latitude"=>0.0, "longitude"=>0.0, "accuracy"=>0, "altitude"=>nil, "altitudeAccuracy"=>nil}, "uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}, "website_token"=>"[FILTERED]", "cw_conversation"=>"eyJhbGciOiJIUzI1NiJ9.eyJzb3VyY2VfaWQiOiI1OTExNjU1Mi1mM2U3LTQ1MjUtYmEyNS1jZDljZDAzM2Y2ODQiLCJpbmJveF9pZCI6MX0.tuf5tICzhLaYYxshOmub1KVsM_oTL09dH5PnS3h4EjU", "conversation"=>{"custom_attributes"=>{"software_version"=>"6.5", "displayMode"=>"browser tab", "languages"=>["en-GB"], "connection"=>{"downlink"=>3.95, "effectiveType"=>"4g", "rtt"=>100}, "userAgent"=>"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36", "userAgentData"=>{"brands"=>[{"brand"=>"Chromium", "version"=>"112"}, {"brand"=>"Google Chrome", "version"=>"112"}, {"brand"=>"Not:A-Brand", "version"=>"99"}], "mobile"=>false, "platform"=>"Windows"}, "ipAddress"=>"", "geo"=>{"latitude"=>0.0, "longitude"=>0.0, "accuracy"=>0, "altitude"=>nil, "altitudeAccuracy"=>nil}, "uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}}}
Apr 19 10:43:54 I, [2023-04-19T05:13:54.532534 #14] INFO -- : [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] Completed 500 Internal Server Error in 17ms (ActiveRecord: 7.0ms | Allocations: 6441)
Apr 19 10:43:54 {"ts":"2023-04-19T05:13:54.534Z","pid":15,"tid":"2evr","lvl":"INFO","msg":"start","ctx":{"class":"ActionCableBroadcastJob","jid":"293a606d9b66bc69aa6b1ab3"}}
Apr 19 10:43:54 I, [2023-04-19T05:13:54.534406 #14] INFO -- : [79cca148-c9ab-4d72-9ea2-986399ab7746] [ActiveJob] Enqueued ActionCableBroadcastJob (Job ID: a8e31e73-325c-4b4b-b87c-019508222e7c) to Sidekiq(default) with arguments: ["UC4fpUBmnVuyL2aB7CR7BkSY", "Dof5nNRp8SjsxqQvzjfXaT2j"], "contact.updated", {:additional_attributes=>{}, :custom_attributes=>{"uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}, :email=>nil, :id=>3330, :identifier=>nil, :name=>"old-shadow-674", :phone_number=>nil, :thumbnail=>"", :type=>"contact", :account_id=>2}
Apr 19 10:43:54 I, [2023-04-19T05:13:54.534608 #15] INFO -- : [ActiveJob] [ActionCableBroadcastJob] [a8e31e73-325c-4b4b-b87c-019508222e7c] Performing ActionCableBroadcastJob (Job ID: a8e31e73-325c-4b4b-b87c-019508222e7c) from Sidekiq(default) enqueued at 2023-04-19T05:13:54Z with arguments: ["UC4fpUBmnVuyL2aB7CR7BkSY", "Dof5nNRp8SjsxqQvzjfXaT2j"], "contact.updated", {:additional_attributes=>{}, :custom_attributes=>{"uuid"=>"08506211-e702-4959-af9d-f4dc5ab51bb7"}, :email=>nil, :id=>3330, :identifier=>nil, :name=>"old-shadow-674", :phone_number=>nil, :thumbnail=>"", :type=>"contact", :account_id=>2}
Apr 19 10:43:54 I, [2023-04-19T05:13:54.535446 #15] INFO -- : [ActiveJob] [ActionCableBroadcastJob] [a8e31e73-325c-4b4b-b87c-019508222e7c] Performed ActionCableBroadcastJob (Job ID: a8e31e73-325c-4b4b-b87c-019508222e7c) from Sidekiq(default) in 0.72ms
Apr 19 10:43:54 {"ts":"2023-04-19T05:13:54.535Z","pid":15,"tid":"2evr","lvl":"INFO","msg":"done","ctx":{"class":"ActionCableBroadcastJob","jid":"293a606d9b66bc69aa6b1ab3","elapsed":0.002}}
Apr 19 10:43:54 {"ts":"2023-04-19T05:13:54.536Z","pid":15,"tid":"2exz","lvl":"INFO","msg":"start","ctx":{"class":"EventDispatcherJob","jid":"aa7190a358c5f797ff1d29a7"}}
Apr 19 10:43:54 F, [2023-04-19T05:13:54.537187 #14] FATAL -- : [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c]
Apr 19 10:43:54 [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] NoMethodError (undefined method `update!' for nil:NilClass
Apr 19 10:43:54
Apr 19 10:43:54 conversation.update!(custom_attributes: permitted_params[:custom_attributes])
Apr 19 10:43:54 ^^^^^^^^):
Apr 19 10:43:54 [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c]
Apr 19 10:43:54 [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] app/controllers/api/v1/widget/conversations_controller.rb:69:in `set_custom_attributes'
Apr 19 10:43:54 [795bcf05-7080-4d7c-baeb-e07f0b4cdc7c] app/controllers/concerns/request_exception_handler.rb:11:in `handle_with_exception'
Apr 19 10:44:15 [b40e44bf-4008-4c3b-a7de-0a34d04e7205] app/controllers/concerns/switch_locale.rb:24:in `set_locale'
Apr 19 10:44:15 [b40e44bf-4008-4c3b-a7de-0a34d04e7205] app/controllers/concerns/switch_locale.rb:11:in `switch_locale'
Apr 19 10:44:15 I, [2023-04-19T05:14:15.483960 #14] INFO -- : source=rack-timeout id=b40e44bf-4008-4c3b-a7de-0a34d04e7205 timeout=15000ms service=17ms state=completed
```

**Reply by scmmishra** [reply]
Thanks, I can take it from here

**Reply by scmmishra** [reply]
@arladmin seems like this happens when the conversation has not started on the website. At the moment you can wait for the `chatwoot:on-message` event before sending the payload. To ensure that a conversation exists.

Meanwhile we will work on a better solution for this

**Comment by sojan-official** [comment]
## 2.16.0 Dev Changelog

### API Endpoints

#### Additions
- `/api/v1/accounts/{account.id}/cache_keys` [GET]
     - https://github.com/chatwoot/chatwoot/pull/6710
     - Endpoint to get cache keys for an account
- `/api/v1/accounts/{account.id}/portals/{portal.slug}/articles/reorder` [POST]
  -  https://github.com/chatwoot/chatwoot/pull/6833
  - the ability to reorder support articles
- `/api/v1/widget/conversations/set_custom_attributes` [POST]
  - https://github.com/chatwoot/chatwoot/pull/6782
  - Allow adding custom attributes to conversations from the widget SDK
- `/api/v1/widget/conversations/destroy_custom_attributes` [POST]
  - https://github.com/chatwoot/chatwoot/pull/6782
  - Allow removing custom attributes to conversations from the widget SDK
- `/super_admin/accounts/{account_id}` [DELETE ]
  - https://github.com/chatwoot/chatwoot/pull/6885/
  - Delete Account from super admin
  
#### Deletions
- none

---

### Other Changes
- `/api/v1/accounts/{account.id}/audit_logs` [GET]
  - https://github.com/chatwoot/chatwoot/pull/6803
  -  Added pagination and additional pagination metadata to the endpoint. Only available in enterprise edition.

**Comment by OllyJack** [comment]
Upgraded to version 2.16 and when replying to livechats, the agents messages do not always go through, and call the following error:

Console error is:

TypeError: Cannot read properties of undefined (reading 'useInboxAvatarForBot')
    at u.useInboxAvatarForBot (configMixin.js:4:44)
    at dn.get (vue.common.prod.js:1445:21)
    at dn.evaluate (vue.common.prod.js:1495:21)
    at u.useInboxAvatarForBot (vue.common.prod.js:1607:32)
    at u.avatarUrl (AgentMessage.vue:133:1)
    at dn.get (vue.common.prod.js:1445:21)
    at dn.evaluate (vue.common.prod.js:1495:21)
    at u.avatarUrl (vue.common.prod.js:1607:32)
    at u.<anonymous> (AgentMessage.vue?ed24:1:403)
    at e._render (vue.common.prod.js:1880:21)


We're using Linux self-hosted.
